cursor_test.go

```diff
@@ -742,6 +745,95 @@ func TestCursor_QuickCheck_BucketsOnly_Reverse(t *testing.T) {
 	}
 }
 
+func TestCursor_DeleteAndPut_InTransactions_Forward(t *testing.T) {
+	testCursorWithBothDeleteAndPut(t,
+		func(c *bolt.Cursor) ([]byte, []byte) {
+			return c.First()
+		}, func(c *bolt.Cursor) ([]byte, []byte) {
+			return c.Next()
+		})
+}
+
+func TestCursor_DeleteAndPut_InTransactions_Reverse(t *testing.T) {
+	testCursorWithBothDeleteAndPut(t,
+		func(c *bolt.Cursor) ([]byte, []byte) {
+			return c.Last()
+		}, func(c *bolt.Cursor) ([]byte, []byte) {
+			return c.Prev()
+		})
+}
+
+func testCursorWithBothDeleteAndPut(t *testing.T, initPos func(*bolt.Cursor) ([]byte, []byte), nextPos func(*bolt.Cursor) ([]byte, []byte)) {
+	testCases := []struct {
+		name        string
+		separateTxn bool
+	}{
+		{
+			name:        "put and delete operations in one transaction",
+			separateTxn: false,
+		},
+		{
+			name:        "put and delete operations in separate transactions",
+			separateTxn: true,
+		},
+	}
+
+	const count = 10
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			db := btesting.MustCreateDB(t)
+
+			err := db.Update(func(tx *bolt.Tx) error {
+				b, _ := tx.CreateBucket([]byte("widgets"))
+				for i := 0; i < count; i++ {
+					k := make([]byte, 8)
+					binary.BigEndian.PutUint64(k, uint64(i))
+					if perr := b.Put(k, make([]byte, 100)); perr != nil {
```

Comment:
suggest adding matric with different value size to construct different stack length for test.

Reply:
Different value size is for separate Txn after rebalancing.

Reply:
IMO, yes. just want to verify it when the p.node is nil

Reply:
Agree that we should add more test, but actually it has bigger concern https://github.com/etcd-io/bbolt/pull/611#issuecomment-1814879792